### PR TITLE
Nexus events cherry pick (reapply) support

### PR DESCRIPTION
**Pull request by bergundy**
## What changed?

Add the `CherryPick` (a better name than "reapply" IMHO) method to the `hsm.EventDefinition` interface and use this method when "reapplying" events from a different branch.

Note that workflow reset with nexus operations is broken right now (and was already broken) as the completion token allocated in the start request targets a specific run. That will be fixed later on.

## Why?

Found an edge case during replication when namespace failover happens as a `NexusOperationCompleted` event is buffered causing the caller workflow to get stuck waiting for completion to arrive.

## How did you test it?

Functional + unit tests

### CODE REVIEW

File: service/history/ndc/workflow_resetter.go [hunk 1]
```diff
@@ -795,8 +798,22 @@ func reapplyEvents(
 			}
 			reappliedEvents = append(reappliedEvents, event)
 		default:
```
Comment (yycptt) [inline service/history/ndc/workflow_resetter.go:800]: cc @yux0 Is the `isDuplicate(event)` check still useful today? I know it's in-memory and doesn't provide any kind of dedup guarantee. Shall we deprecated that or there's a plan for making it real?

**Review by yycptt — APPROVED** [review]

File: service/history/ndc/workflow_resetter.go (hunk 1 above)
Comment (yux0) [reply service/history/ndc/workflow_resetter.go:800]: yes. this is best effort. We should dedup them based on the version histories

File: service/history/ndc/workflow_resetter.go [hunk 2]
```diff
@@ -795,8 +798,22 @@ func reapplyEvents(
...
+				// Only reapply hardcoded events above or ones registered and are cherry-pickable in the HSM framework.
+				continue
+			}
+			if err := def.CherryPick(root, event); err != nil {
+				if errors.Is(err, hsm.ErrNotCherryPickable) || errors.Is(err, hsm.ErrStateMachineNotFound) || errors.Is(err, hsm.ErrInvalidTransition) {
+					continue
+				}
+				return reappliedEvents, err
+			}
+			mutableState.AddHistoryEvent(event.EventType, func(he *historypb.HistoryEvent) {
```
[the remark this answers is not in the document]
Comment (bergundy) [reply service/history/ndc/workflow_resetter.go:813]: This will actually record the history event, otherwise we'll have applied the transition without the corresponding event.

[the remark this answers is not in the document]
Comment (yux0) [reply service/history/ndc/workflow_resetter.go:813]: I understand it now

**Comment by bergundy** [comment]
Tests passed here but the status isn't being updated for some reason. looks like a GH bug.

https://github.com/temporalio/temporal/actions/runs/9896329597/job/27376679898?pr=6263.

I'll try to close and reopen.
